ee/clickhouse/sql/plugin_log_entries.py

```diff
@@ -0,0 +1,66 @@
+from ee.kafka_client.topics import KAFKA_PLUGIN_LOG_ENTRIES
+
+from .clickhouse import KAFKA_COLUMNS, kafka_engine, table_engine, ttl_period
+
+PLUGIN_LOG_ENTRIES_TABLE = "plugin_log_entries"
+
+PLUGIN_LOG_ENTRIES_TABLE_BASE_SQL = """
+CREATE TABLE {table_name}
+(
+    id UUID,
+    team_id Int64,
+    plugin_id Int64,
+    plugin_config_id Int64,
+    timestamp DateTime64(6, 'UTC'),
+    type VARCHAR,
+    message VARCHAR,
+    instance_id UUID
+    {extra_fields}
+) ENGINE = {engine}
+"""
+
+PLUGIN_LOG_ENTRIES_TABLE_SQL = (
+    PLUGIN_LOG_ENTRIES_TABLE_BASE_SQL
+    + """PARTITION BY toYYYYMMDD(timestamp)
+ORDER BY (plugin_config_id, toHour(timestamp), timestamp, id)
+{ttl_period}
+SETTINGS index_granularity=512
```

Comment:
I'm not much of a CH wizard – are these parameters optimal for the logs use case, where most selects will be by `plugin_config_id` and `timestamp`? @fuziontech

Reply:
🤔 
Considering that we'll be TTL-ing logs for 30 days and 99.999% of the time filtering by `plugin_config_id` I would partition and sort it like this:
```sql
PARTITION BY plugin_config_id
ORDER BY (timstamp, id)
```

This makes reads and writes to the parts very straightforward.

Reply:
Thanks, this is helpful! Though I think this could result in a bit more partitions than needed – we should keep the number of them under 1k per table, right? And `plugin_config_id` for most intents and purposes is a `(team_id, plugin_id)` pair, so once we start preinstalling plugins (like GeoIP) such a schema would result in a partition per plugin per team, so sooner or later we'd exceed 1k. Perhaps:
```
PARTITION BY plugin_id
ORDER BY (team_id, id)
```
(since UUIDT is time-sortable too)

Reply:
What is `index_granularity` about BTW?

Reply:
`index_granularity` is basically the max size in rows between marks in the index. Part of the reason why clickhouse is so powerful is what it does on disk. https://clickhouse.tech/docs/en/engines/table-engines/mergetree-family/mergetree/#mergetree-data-storage

Having a ton fo partitions is not ideal, but only from an OLAP POV. It slows things down if you need to grab a ton of partitions because you have to open more files. In this case you will aways b going to one partition and then grabbing things out of it in a super reliable ordered way. Having too many partitions shouldn't bet too much of an issue.